### ISSUE TITLE
espeak does not connect gen case in Farsi / Persian with E

### ISSUE THREAD

**Issue by hkalbasi** [espeak-ng/espeak-ng#862]
This is a huge problem with espeak Farsi. For example: Physic laws are beautiful.
```
قوانین فیزیک جالب هستند
```
espeak say it:
```
ghavaanin fizik jaleb hastand
```
but the correct is:
```
ghavaanine fizik jaleb hastand
```
notice the "E". This problem is very common, I can say in every sentence we have this E connections and this means espeak reads every sentence wrong :)

[hazm library](https://github.com/sobhe/hazm) can detect this but it is based on machine learning and needs a 20 megabytes model. Can we use this in espeak-ng? For example by making Farsi a seperate installable package?

**Comment by jaacoppi** [comment]
Is it possible to input the text to hazm and get the correct text as output, and then pipe that to espeak-ng? If yes, you could try a script that pipes everything through hazm first. Something like hazm "text" | espeak-ng -v fa

I don't know if combining hazm and espeak-ng is simple because hazm is a python script.

**Comment by HKalbasi** [comment]
Yes it is possible but a built-in solution inside espeak-ng would be way more useful. Very tools, like screen readers have out of box support for espeak-ng but it is not easy to use hazm in them in this way.

**Comment by hkalbasi** [comment]
I guess espeak is not relevant anymore in this LLM era, given that 30M param models can outcompete it? Closing due inactivity anyway.

**Comment by alex19EP** [comment]
No hard feelings about closing it, but let me leave a note for anyone who lands here later — ezafe comes up regularly.

**Status today (master, 2026-08).** Unchanged: espeak-ng does not *predict* ezafe on undiacritized text.

```
$ espeak-ng -xq -v fa "قوانین فیزیک جالب هستند"
q1av'Anin f'izik dZ'Aleb h'astand
```

**What does work.** If the kasre is written explicitly, espeak-ng honours it:

```
$ espeak-ng -xq -v fa "قوانینِ فیزیک جالب هستند"
q1av'Anin,e f'izik dZ'Aleb h'astand

$ espeak-ng -xq -v fa "کتابِ من"
ket'Abe m'an

$ espeak-ng -xq -v fa "خانهٔ من"
x'Ane:j,e m'an
```

So the split you suggested already exists in a workable form: anything that can restore diacritics (hazm, or your own normaliser) can feed espeak-ng and get the right output, with no change on our side. That is also how the screen reader case can be solved — a text filter in front of the synthesiser, rather than inside it.

**Why the model won't be bundled.** Ezafe placement needs POS/syntax-level information; `dictsource/fa_rules` is a letter-to-phoneme rule engine and structurally can't express it. And espeak-ng's constraints are dependency-free C, a few MB of data for 100+ languages, and deterministic offline output on hardware where a 20 MB per-language model isn't affordable. Making `fa` a separately installable package with an ML runtime dependency would be a first for the project, and it isn't something I'd take on.

What *would* be reviewed: rule-level improvements in `dictsource/fa_rules` / `fa_list` for high-frequency constructions where ezafe is predictable from the surface form alone. That won't cover noun–noun ezafe, but it's the tractable part.

**On "espeak is not relevant in the LLM era".** Different niche rather than a losing one. The users are mostly screen reader users (NVDA, Orca, TalkBack), where what matters is very low latency, intelligibility at high speech rates, no GPU, no network, and 100+ languages in a few megabytes — a 30M-param neural model doesn't fit that budget on a low-end Android phone or a Raspberry Pi. And the neural systems themselves lean on espeak-ng: Piper and several Kokoro/StyleTTS pipelines call it for G2P, which is precisely the ezafe problem above showing up again one layer higher.

Leaving this closed as not planned. If anyone wants to attempt the rule-level part, a PR against `dictsource/fa_rules` is welcome.
